### PR TITLE
add test vectors for address resolution using chaos actor

### PR DESCRIPTION
**Pull request by willscott**
address #47

### CODE REVIEW

File: gen/builders/predicates.go [hunk 1]
```diff
@@ -33,6 +34,17 @@ func ExitCode(expect exitcode.ExitCode) ApplyRetPredicate {
 	}
 }
 
+// MessageReturns returns an ApplyRetPredicate that passes if the message response
+// matches the argument.
+func MessageReturns(expect []byte) ApplyRetPredicate {
```
Comment (raulk) [inline gen/builders/predicates.go:39]: This would be more concise if the arg was a `CBORUnmarshaller`. WDYT? Rationale: if we're comparing against something that has indeed been returned, we know it must be CBOR.

File: gen/suites/vm_violations/address_resolution.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	alice := v.Actors.Account(address.SECP256K1, abi.NewTokenAmount(1_000_000_000_000))
+	v.CommitPreconditions()
+
+	v.Messages.Raw(alice.ID, chaos.Address, chaos.MethodResolveAddress, MustSerialize(&builtin.SystemActorAddr), Nonce(0), Value(big.Zero()))
+
+	invalidIDAddr, _ := address.NewIDAddress(77)
+	v.Messages.Raw(alice.ID, chaos.Address, chaos.MethodResolveAddress, MustSerialize(&invalidIDAddr), Nonce(1), Value(big.Zero()))
+	v.CommitApplies()
+
+	v.Assert.EveryMessageResultSatisfies(ExitCode(exitcode.Ok))
```
Comment (raulk) [inline gen/suites/vm_violations/address_resolution.go:26]: Check that the response contains the input identity address?

File: gen/suites/vm_violations/address_resolution.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+	invalidActorAddr, _ := address.NewActorAddress([]byte("invalid"))
+	v.Messages.Raw(alice.ID, chaos.Address, chaos.MethodResolveAddress, MustSerialize(&invalidActorAddr), Nonce(0), Value(big.Zero()))
+	v.CommitApplies()
+
+	v.Assert.EveryMessageResultSatisfies(ExitCode(exitcode.Ok))
+	v.Assert.EveryMessageResultSatisfies(MessageReturns(&builtin.SystemActorAddr))
+}
+
+func actorResolutionExistant(v *Builder) {
```
Comment (raulk) [inline gen/suites/vm_violations/address_resolution.go:43]: This test is actually testing a different case than:

> If the {pubkey|actor} address is in the InitActor's table, the corresponding ID address should be returned

This test is not using an actor address, but a SECP256K1 address. Suggest you create a payment channel actor, and use that address in the resolution.

File: chaos/actor.go [hunk 4]
```diff
@@ -115,3 +117,19 @@ func (a Actor) CreateActor(rt runtime.Runtime, args *CreateActorArgs) *adt.Empty
...
+	Address address.Address
+	Success bool
+}
+
+func (a Actor) ResolveAddress(rt runtime.Runtime, args *address.Address) *address.Address {
+	rt.ValidateImmediateCallerAcceptAny()
+	
+	resolvedAddr, ok := rt.ResolveAddress(*args)
+	if !ok {
+		invalidAddr, _ := address.NewIDAddress(0)
```
Comment (raulk) [inline chaos/actor.go:131]: Using the system actor's implicit address as a sentinel value for when resolution fails is hard to follow. How about we add a `ResolveAddressReturn` type that contains the address and a boolean indicating if resolution failed? It's a lot more explicit and self-descriptive. See how `CreateActor` uses this technique on the input side.

**Review by raulk — CHANGES_REQUESTED** [review]

File: chaos/actor.go (hunk 4 above)
Comment (willscott) [reply chaos/actor.go:131]: the downsides of new types is that i would also need to either add cbor-gen machinery around it, as would any other implementations, right?

Comment (raulk) [reply chaos/actor.go:131]: Good news is that it's already there ;-) Add your type to: https://github.com/filecoin-project/test-vectors/blob/master/chaos/gen/gen.go, then execute `go generate`

https://github.com/filecoin-project/test-vectors/blob/a5d1c62f093bd44e487a35211031a5e5a0824936/chaos/actor.go#L13

File: gen/suites/vm_violations/address_resolution.go (hunk 3 above)
Comment (willscott) [reply gen/suites/vm_violations/address_resolution.go:43]: Other tests will implicitly check that type as well, right? i guess more types in this test can help localize vm misbehavior

File: gen/suites/vm_violations/address_resolution.go (hunk 2 above)
Comment (willscott) [reply gen/suites/vm_violations/address_resolution.go:26]: The predicate only works on test vectors that make a single message. for these, we expect different outputs for the 2 messages. worth splitting into different tests?

Comment (willscott) [reply gen/suites/vm_violations/address_resolution.go:26]: (i'm less worried about the asserts doing 'full validation', because the actual validation when later running test vectors will just be exact equivalency of receipts as i understand it, while these asserts are more there to ensure is sane, and are lotus specific. We believe the spec-actors implementation used by lotus to already be tested, so spending a bunch of time on predicates / assertions to test that at generation time is less useful than creating the vectors with receipts that can be used by other actors implementations to test their implementations. maybe that's not the right way to be thinking about this, though?)

Comment (raulk) [reply gen/suites/vm_violations/address_resolution.go:26]: So these assertions are really sanity checks. I kinda understand that because we've implemented the Chaos Actor behaviour that we're triggering now, we have a good sense that it's working as intended, but it's still worth doing these asserts for pragmatism.

The receipt comparison helps us pinpoint changes, but if the second message happens to be returning `t88` instead of `t77` (for whatever reason we didn't anticipate), we would've never caught it.

I really think we should add those assertions here.
